package/sbp_anpp_bridge/src/sbp_anpp_bridge.c

```diff
@@ -50,29 +51,67 @@
 #define SBP_MSG_ODOMETRY_FLAGS_SRC2 (0x08u)
 #define SBP_MSG_ODOMETRY_FLAGS_SRC3 (0x0Cu)
 
-bool anpp_debug = false;
+#define INVALID_TOW 0xffffffff
+#define TIME_SOURCE_MASK 0x07 /* Bits 0-2 */
+#define NO_TIME 0
+#define MS_IN_WEEK 7 * 24 * 60 * 60 * 1000
 
+bool anpp_debug = false;
+unsigned int last_gpstime_tow = INVALID_TOW;
+struct timespec last_gps_systime;
+int odo_time_offset_ms = 0;
 pk_endpoint_t *anpp_pub = NULL;
 
-static u32 get_time_ms(void)
+static void gps_time_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  (void)context;
+  (void)sender_id;
+  (void)len;
+  msg_gps_time_t *time = (msg_gps_time_t *)msg;
+
+  if ((time->flags & TIME_SOURCE_MASK) == NO_TIME) {
+    return;
+  }
+
+  clock_gettime(CLOCK_MONOTONIC, &last_gps_systime);
+  last_gpstime_tow = time->tow;
+}
+
+static int notify_time_offset_changed(void *context)
 {
-  struct timespec now;
-  clock_gettime(CLOCK_MONOTONIC, &now);
-  return (now.tv_sec * 1000 + now.tv_nsec / 1000000);
+  (void)context;
+  return 0;
 }
 
 static void convert_anpp_odometer_to_sbp(odometer_packet_t *an_odo_packet, msg_odometry_t *odo_msg)
 {
-  odo_msg->tow = (u32)get_time_ms();
-  odo_msg->velocity = (s32)M_TO_MM(an_odo_packet->speed);
-  odo_msg->flags = (u8)(SBP_MSG_ODOMETRY_FLAGS_SRC0 | SBP_MSG_ODOMETRY_FLAGS_SYSTIME);
+  if (last_gpstime_tow != INVALID_TOW) {
+    double tow = last_gpstime_tow;
+    struct timespec now;
+    clock_gettime(CLOCK_MONOTONIC, &now);
+    /* Add on elapsed time since gps_time_callback was received */
+    tow += (now.tv_sec * 1000.0 + now.tv_nsec / 1000000.0)
+           - (last_gps_systime.tv_sec * 1000.0 + last_gps_systime.tv_nsec / 1000000.0);
+    tow += odo_time_offset_ms;
+    if (tow > MS_IN_WEEK) {
+      tow -= MS_IN_WEEK;
+    }
+    if (tow < 0) {
```

Comment:
Just curious which case this is catching, if tow rolls over between the last GPS time and odo sample? Still not super familiar with tow rollover behavior

Reply:
This is catching if my addition / subtraction for the time puts is into the previous or the next week.  If the addend I added to the time of week has more than the number of seconds in the week, then I need to wrap around zero.  If it is negative, I need to add to it.

Reply:
This is to ensure that the time of week in our output odo message is between 0 and MS_IN_WEEK.  After I add or subtract from the incoming time of week based upon time of receipt, I could make a negative time of week or one that has more than MS_IN_WEEK.  I think these statements guard against that.